### PR TITLE
add `google_iam_principal_access_boundary_policy` resource

### PR DESCRIPTION
**Pull request by derekchu-google**
Fixes https://github.com/hashicorp/terraform-provider-google/issues/19905

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_iam_principal_access_boundary_policy` (beta)
```

### CODE REVIEW

**Comment by derekchu-google** [comment]
still WIP. please hold off on review. thanks!

**Comment by derekchu-google** [comment]
> 🟢 Tests passed during RECORDING mode: `TestAccBigqueryConnectionConnection_bigqueryConnectionKmsExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12044/artifacts/60fbf356-59e1-4a3a-8729-8aa22741c43f/recording/TestAccBigqueryConnectionConnection_bigqueryConnectionKmsExample.log)] `TestAccDataLossPreventionDiscoveryConfig_Update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12044/artifacts/60fbf356-59e1-4a3a-8729-8aa22741c43f/recording/TestAccDataLossPreventionDiscoveryConfig_Update.log)]
> 
> 🟢 No issues found for passed tests after REPLAYING rerun.
> 
> 🔴 Tests failed during RECORDING mode: `TestAccIAM3PrincipalAccessBoundaryPolicy_iam3PrincipalAccessBoundaryPolicyExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12044/artifacts/60fbf356-59e1-4a3a-8729-8aa22741c43f/build-log/recording_build/TestAccIAM3PrincipalAccessBoundaryPolicy_iam3PrincipalAccessBoundaryPolicyExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12044/artifacts/60fbf356-59e1-4a3a-8729-8aa22741c43f/recording/TestAccIAM3PrincipalAccessBoundaryPolicy_iam3PrincipalAccessBoundaryPolicyExample.log)] `TestAccIAM3PrincipalAccessBoundaryPolicy_iam3PrincipalAccessBoundaryPolicyExample_update`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12044/artifacts/60fbf356-59e1-4a3a-8729-8aa22741c43f/build-log/recording_build/TestAccIAM3PrincipalAccessBoundaryPolicy_iam3PrincipalAccessBoundaryPolicyExample_update_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12044/artifacts/60fbf356-59e1-4a3a-8729-8aa22741c43f/recording/TestAccIAM3PrincipalAccessBoundaryPolicy_iam3PrincipalAccessBoundaryPolicyExample_update.log)]
> 
> 🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.
> 
> View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12044/artifacts/60fbf356-59e1-4a3a-8729-8aa22741c43f/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12044/artifacts/60fbf356-59e1-4a3a-8729-8aa22741c43f/recording) for each test

i tested with my own project/organization. this command is hidden behind a visibility label so I impersonated with a service account which has the proper permissions. im not sure how to reflect this testing setup in these PR tests.

**Comment by derekchu-google** [comment]
> Fixes [hashicorp/terraform-provider-google#19905](https://github.com/hashicorp/terraform-provider-google/issues/19905)
> 
> **Release Note Template for Downstream PRs (will be copied)**
> 
> ```
> `google_iam3_principal_access_boundary_policy`
> ```

Thank yo

**Comment by derekchu-google** [comment]
sorry hit the wrong button and closed it by accident

File: mmv1/third_party/terraform/.teamcity/components/inputs/services_ga.kt
Comment (SarahFrench) [inline mmv1/third_party/terraform/.teamcity/components/inputs/services_ga.kt:443]: This is a consequence of making the new resource (and the wider service) beta-only.

```suggestion
```

File: mmv1/products/iam3/product.yaml
Comment (SarahFrench) [inline mmv1/products/iam3/product.yaml:18]: I can only see the [a beta 'v3beta' version of iam3](https://cloud.google.com/iam/docs/reference/rest/v3beta/organizations.locations.principalAccessBoundaryPolicies) in the public docs, so I think we need to ensure this new resource uses those endpoints and is only added to the google-beta provider. 

```suggestion
  - name: 'beta'
    base_url: 'https://iam.googleapis.com/v3beta/'
```

File: mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml
Comment (SarahFrench) [inline mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:16]:
```suggestion
description: An IAM principal access boundary policy resource.
```

Copied from https://cloud.google.com/iam/docs/reference/rest/v3beta/organizations.locations.principalAccessBoundaryPolicies

Comment (SarahFrench) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:16]: Could you please also add these fields to the file, where the capitalised text needs to be replaced:

```
references:
  guides:
   # Link to quickstart in the API's Guides section. For example:
   # 'Create and connect to a database': 'https://cloud.google.com/alloydb/docs/quickstart/create-and-connect'
    'QUICKSTART_TITLE': 'QUICKSTART_URL'
  # Link to the REST API reference for the resource. For example,
  # https://cloud.google.com/alloydb/docs/reference/rest/v1/projects.locations.backups
  api: 'API_REFERENCE_URL'
```

They're described in [this section](https://googlecloudplatform.github.io/magic-modules/develop/resource/#add-a-resource) of the contribution guide and helps users find GCP documentation from the resource's provider documentation page.

File: mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml
Comment (SarahFrench) [inline mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:17]:
```suggestion
min_version: 'beta'
id_format: 'organizations/{{organization}}/locations/{{location}}/principalAccessBoundaryPolicies/{{principal_access_boundary_policy_id}}'
```

Adding `min_version: 'beta'` here will ensure this resource is only in google-beta

**Review by SarahFrench — COMMENTED** [review]
👋🏻 Hi, thanks for your PR!

From looking at the API documentation I believe this resource would be in the google-beta provider but not the google provider, so I've made some suggested changes based on that. Is my understanding correct?

**Comment by derekchu-google** [comment]
> 👋🏻 Hi, thanks for your PR!
> 
> From looking at the API documentation I believe this resource would be in the google-beta provider but not the google provider, so I've made some suggested changes based on that. Is my understanding correct?

hi Sarah! thanks for the comments. this resource today is in public preview so yes, you're correct it's a v3beta which would then be google-beta. that being said, terraform support is only targeted for our GA so that's why we're not doing a google-beta. the tricky part is for this PR to get approved and wait until a week before we remove the visibility restriction.


that being said, i do wonder if we should just push a google-beta out first and have that available....

File: mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml
Comment (derekchu-google) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:22]: done

**Comment by derekchu-google** [comment]
@SarahFrench the errors in the tests are unrelated to my changes. this PR should be ready to go

**Comment by SarahFrench** [comment]
Thanks for pivoting to adding the resource to google-beta - once this PR is merged it'll be a few small code changes to promote the resource to GA, and that PR would be easier to coordinate with a GA release.

File: mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml
Comment (SarahFrench) [inline mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:94]: I believe that the etag field would only ever be used internally in the provider, and a user wouldn't set this field in their Terraform configuration. Because of that, I think this field should be `output: true`, instead of `default_from_api: true`. What do you think?

**Review by SarahFrench — COMMENTED** [review]
Here are some comments about a variety of things, but nothing major. The biggest thing is that I think the etag field will need some changes to avoid users trying to set the value.

File: mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml
Comment (derekchu-google) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:97]: @SarahFrench is this for every property or just etag?

File: mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml
Comment (derekchu-google) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:94]: a user can set an etag when calling update so they can ensure they specify a specific resource to update. i added the default_from_api b/c i kept getting an error in the update test for not testing that field.

File: mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml
Comment (derekchu-google) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:154]: good call

File: mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml
Comment (derekchu-google) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:55]: good catch. thanks.

File: mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml
Comment (derekchu-google) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:94]: here's the error from testing:

=== NAME  TestAccIAM3PrincipalAccessBoundaryPolicy_iam3PrincipalAccessBoundaryPolicyExample
    resource_iam3_principal_access_boundary_policy_generated_test.go:42: Step 1/2 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # google_iam3_principal_access_boundary_policy.my-pab-policy will be updated in-place
          ~ resource "google_iam3_principal_access_boundary_policy" "my-pab-policy" {
              - etag                                = "W/\"0MJpU5Y4jPJAMvKXTwD1ig==\"" -> null

Comment (SarahFrench) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:92]: Hmm, that error's interesting.

Is the etag used to prevent concurrent writes to the policy?

Comment (derekchu-google) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:92]: it's moreso something to manage concurrent writes to the policy. by specifying the etag, the user can be absolutely sure of delta of the change.

Comment (SarahFrench) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:92]: Indeed, then I think we'd need to follow the way etags are used elsewhere in the provider. For example [google_organization_policy](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_organization_policy#etag) uses an etag but it's not expected for the user to provide a new etag value each time there's an update. Similarly all IAM resources use etags ([example](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/api_gateway_api_iam#etag)) and they are managed entirely in the provider without user input.

Comment (derekchu-google) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:92]: thanks Sarah. went ahead and followed those resources and set output: true.

**Comment by derekchu-google** [comment]
@SarahFrench thanks for all the reviews! please do let me know if there's additional concerns.

**Comment by SarahFrench** [comment]
Re-running tests, as TestAccContainerCluster_resourceManagerTags appears to have failed due to a bootstrapping issue


/gcbrun

File: mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml
Comment (SarahFrench) [reply mmv1/products/iam3/PrincipalAccessBoundaryPolicy.yaml:92]: Thanks! In the debug logs from the acceptance test I can see that the etag is stored in state after it's received from the API, and in subsequent API calls to update the policy I can see the etag is sent with the update request. The new etag returned from the API is then stored in the resource state and the process repeats. When Terraform performs refresh actions, prior to planning/applying changes, the refreshes will pull in etag changes that may have happened out-of-band.

**Comment by derekchu-google** [comment]
thanks @SarahFrench. interesting when i ran my tests locally on the beta provider it complained about it missing from the ga provider. not sure why that was the case. anyways, appreciate you removing it if unnecessary.

**Comment by SarahFrench** [comment]
>thanks @SarahFrench. interesting when i ran my tests locally on the beta provider it complained about it missing from the ga provider. not sure why that was the case. anyways, appreciate you removing it if unnecessary.

Looks like I shouldn't have done that, as an empty file google/services/iam3/resource_iam_principal_access_boundary_policy_test.go is made in the GA provider. I'll push a change to re-add it

**Review by SarahFrench — APPROVED** [review]
LGTM

**Pull request merged**
